### PR TITLE
fix(deriver): filter non-prose-tagged messages from representation extraction

### PR DESCRIPTION
**Pull request by 24601**
## Summary

Adds a `metadata.type` discriminator to the deriver so messages tagged as **not the observed peer's authored prose** are excluded from the LLM-facing prompt. This prevents pasted diffs, tool-result blocks, and assistant tool actions from producing peer-attributed conclusions in the representation layer.

Companion to plastic-labs/claude-honcho#34, which tags such messages on upload. Either side is independently useful; together they structurally prevent the misattribution failure mode.

## The bug, concretely

The Anthropic Messages API uses `role: "user"` for both the user's authored prose AND for `tool_result` blocks (and any content the user pastes back into a prompt). Today the deriver treats every `role: "user"` message as authored prose, which produces durable `<user> changed/wrote/added X` facts when:

1. **Pasted diffs.** A user prompt like *"You are performing an adversarial code review. Review the provided git diff: ```diff +function buildOperatorPlan(...) ... ``` "*. The diff body rides `role: "user"` per the Messages API; the extractor reads `+`/`-` lines as user statements about their own code and concludes `<user> changed buildOperatorPlan`. Observed in the wild — the conclusion was used as load-bearing context across sessions.

2. **Cross-observation in directional mode.** `aiPeer.message("[Tool] Edited middleware.ts")` is correct on the aiPeer side, but in `directional` observation mode the deriver folds aiPeer messages into the user-peer's representation as `<user> did X`. `unified` mode (the 0.2.4 default) reduces blast radius but does not eliminate it — directional remains a supported config.

In one heavily-used workspace we observed ~12K of ~60K conclusions (~20%) match the misattribution pattern.

## Fix

```python
NON_PROSE_METADATA_TYPES: frozenset[str] = frozenset(
    {
        "user_paste_not_speech",
        "tool_action",
    }
)


def _is_extraction_eligible(msg: Message) -> bool:
    metadata = msg.h_metadata or {}
    return metadata.get("type") not in NON_PROSE_METADATA_TYPES
```

Inside `process_representation_tasks_batch`:

```python
eligible_messages = [m for m in messages if _is_extraction_eligible(m)]
if not eligible_messages:
    logger.debug(
        "All %d messages in batch tagged as non-prose; skipping representation extraction "
        "(observed=%s, latest_message_id=%s)",
        len(messages), observed, latest_message.id,
    )
    return

formatted_messages = "\n".join(
    format_new_turn_with_timestamp(msg.content, msg.created_at, msg.peer_name)
    for msg in eligible_messages
)
```

`earliest_message` / `latest_message` tracking uses the original `messages` list so progress accounting stays correct. Token accounting now sums only eligible messages (we don't bill against tokens we elected to filter).

## Tests

`tests/deriver/test_metadata_filter.py` adds 9 tests mirroring honcho's existing convention (`pytest` async + `Mock` + `patch honcho_llm_call`):

```
TestExtractionEligibility::test_default_metadata_is_eligible
TestExtractionEligibility::test_empty_metadata_is_eligible
TestExtractionEligibility::test_unrelated_metadata_type_is_eligible
TestExtractionEligibility::test_user_paste_not_speech_is_filtered
TestExtractionEligibility::test_tool_action_is_filtered
TestExtractionEligibility::test_non_prose_set_contains_known_tags
TestProcessRepresentationFiltering::test_skips_llm_call_when_all_messages_non_prose
TestProcessRepresentationFiltering::test_filters_non_prose_from_llm_prompt
TestProcessRepresentationFiltering::test_passes_prose_messages_unchanged
```

```bash
$ uv run pytest tests/deriver/test_metadata_filter.py -q
9 passed in 16.14s
```

## Backwards compatibility

- **Default-True predicate** means existing clients that don't tag messages see no behavior change. Tags are opt-in; the world without them is exactly what it is today.
- **Messages stay in the database.** Only the deriver's representation extraction is suppressed for tagged messages. Replay, audit, message-stream completeness, and search are unaffected.
- **Tag set is private to the deriver.** Adding new tags is a one-line addition to `NON_PROSE_METADATA_TYPES`; no API schema change.

## Notes for reviewers

- The set lives in `deriver.py` for proximity to its only consumer. Happy to move it to `src/utils/` or `src/schemas/` if you'd prefer the tag vocabulary to be more discoverable for future filter sites (e.g. dialectic, dream).
- The early-return when every message in a batch is non-prose is the only "shape" change to the extractor; the per-message filter in the formatted-prompt comprehension is otherwise the minimal possible change.
- I did not touch the dream / dialectic layers. Those are downstream consumers of the same message stream and likely benefit from the same filter, but they're separate code paths and merit their own PR with their own tests.
- If you'd prefer the `type` key be namespaced (e.g. `honcho.role` / `honcho.subject`), I'm happy to rename. The plugin PR uses bare `type` because that's the existing convention in `session-end.ts` (`type: "assistant_prose"`, `type: "assistant_brief"`).

## Summary by CodeRabbit

* **New Features**
  * Metadata-driven filtering: messages marked non-prose are excluded from LLM prompts and token accounting; if no prose remains the LLM call is skipped. Persisted provenance and attributions now reference only prose-eligible source messages.

* **Tests**
  * Added unit and integration tests for metadata eligibility, mixed batches, edge cases, LLM-skip behavior, prompt construction, and provenance/attribution filtering.

### CODE REVIEW

File: src/deriver/deriver.py [hunk 1]
```diff
@@ -59,6 +94,20 @@ async def process_representation_tasks_batch(
...
+    eligible_messages = [m for m in messages if _is_extraction_eligible(m)]
+    if not eligible_messages:
+        logger.debug(
+            "All %d messages in batch tagged as non-prose; skipping representation extraction "
+            "(observed=%s, latest_message_id=%s)",
+            len(messages),
+            observed,
+            latest_message.id,
+        )
+        return
```
[the remark this answers is not in the document]
Comment (24601) [reply src/deriver/deriver.py]: Addressed in `2547842`. The early-return path now emits a minimal `RepresentationCompletedEvent` (`explicit_conclusion_count=0`, `input_tokens=0`, `output_tokens=0`, `message_count=len(messages)`, `queue_items_processed=len(queue_item_message_ids)`) and records `starting_message_id` / `ending_message_id` / `total_processing_time` / `observation_count` metrics + `log_performance_metrics` before returning. Same shape as the normal completion path, so dashboards summing these counters see every batch outcome — including all-non-prose ones — uniformly. New test `test_emits_minimal_completion_event_on_early_return` patches `src.deriver.deriver.emit` and asserts the event is emitted exactly once with the right zero-values; 15/15 pass.

**Comment by VVoruganti** [comment]
The appropriate way to do this with the API is to toggle reasoning on that specific message using the api `MessageConfiguration` model: https://docs.honcho.dev/v3/api-reference/endpoint/messages/create-messages-for-session#body-messages-items-configuration-one-of-0-reasoning-one-of-0
